[macOS] Web content process may crash when context menu is invoked

lilyspiniolas opened the pull request:
#### 9a322f4ab7cee3a527f4df4f1dccf1a9050e9296
<pre>
[macOS] Web content process may crash when context menu is invoked
<a href="https://bugs.webkit.org/show_bug.cgi?id=293342">https://bugs.webkit.org/show_bug.cgi?id=293342</a>
<a href="https://rdar.apple.com/151711558">rdar://151711558</a>

Reviewed by Abrar Rahman Protyasha.

Prior to these changes, if the context menu was invoked and the hit node
was in the shadow DOM, was not an element, and had no parent node, the web
content process would crash. This is due to an error in `prepareContextForQRCode`,
where we would attempt to downcast the node into an element. If this failed, we
attempted to get the node&apos;s parent element instead. In the scenario listed above,
the node is not an element and had no parent, so when we dereferenced the
non-existent parent, a crash occurred.

These changes resolve this issue by checking the lineage of
`parentOrShadowHostElement` instead of `protectedParentElement`, and ensures
it is non-null before dereferencing.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::prepareContextForQRCode):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuTopLevelTextNodeInShadowDOMShouldNotCrash)):

Canonical link: <a href="https://commits.webkit.org/295316@main">https://commits.webkit.org/295316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/847e5e63dde2eea7ba028665215845140909f5eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104602 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109814 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55273 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106642 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24699 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32857 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79427 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94402 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18991 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12477 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54646 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88702 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112205 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88512 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32127 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90636 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88131 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22482 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33034 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10800 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27080 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31690 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37032 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31483 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34821 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33042 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
I think this may be testable!

aprotyas reviewed (approved):
r+ but a test would be nice if possible

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
```diff
@@ -870,6 +870,27 @@ static void swizzledOrderFrontColorPanel(id, SEL, id)
+        shadowRoot.appendChild(textNode);
+    </script>
+    )")];
+    elementInfo = [webView rightClickAtPointAndWaitForContextMenu:NSMakePoint(5, 395)];
```
Nit: just declare here. Also lets you CTAD
```suggestion
    RetainPtr elementInfo = [webView rightClickAtPointAndWaitForContextMenu:NSMakePoint(5, 395)];
```

lilyspiniolas committed:
[macOS] Web content process may crash when context menu is invoked
https://bugs.webkit.org/show_bug.cgi?id=293342
rdar://151711558

Reviewed by Abrar Rahman Protyasha.

Prior to these changes, if the context menu was invoked and the hit node
was in the shadow DOM, was not an element, and had no parent node, the web
content process would crash. This is due to an error in `prepareContextForQRCode`,
where we would attempt to downcast the node into an element. If this failed, we
attempted to get the node's parent element instead. In the scenario listed above,
the node is not an element and had no parent, so when we dereferenced the
non-existent parent, a crash occurred.

These changes resolve this issue by checking the lineage of
`parentOrShadowHostElement` instead of `protectedParentElement`, and ensures
it is non-null before dereferencing.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::prepareContextForQRCode):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST(ContextMenuTests, ContextMenuTopLevelTextNodeInShadowDOMShouldNotCrash)):

Canonical link: https://commits.webkit.org/295316@main
